chef/cookbooks/crowbar/recipes/crowbar-db-dump.rb

```diff
@@ -0,0 +1,87 @@
+#
+# Cookbook Name:: crowbar
+# Recipe:: crowbar-db-dump
+#
+# Copyright 2013-2016, SUSE LINUX Products GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+#
+# This recipe is for executing actions that need to be done at nodes
+# as a preparation for the system upgrade. More specifically, this is executed
+# for nodes that have database installed because they need to be handled differently
+# than non DB nodes.
+#
+# First the database content should be dumped. In the next run (indicated by extra argument),
+# the database needs to be shut down.
+
+return unless node[:platform_family] == "suse"
+
+if node["crowbar_wall"]["crowbar_openstack_shutdown"]
+
+  include_recipe "crowbar::stop-services-before-upgrade"
+
+  # If postgres is not running here, it means we're in the DB cluster and database runs
+  # on the other node: let the other node take care of the rest.
+  return unless ::Kernel.system("service postgresql status")
+
+  # Check the available space before the dump
+
+  dump_location = node[:crowbar][:upgrade][:db_dump_location]
+
+  ruby_block "check available space" do
+    block do
+      require "pathname"
+      path = Pathname.new(dump_location).split.first.to_s
+      available = `df #{path} | tail -n 1 | sed "s/ \\+/ /g" | cut -d " " -f 4`
+      db_size = `su - postgres -c 'pg_dumpall | wc -c'`
+      if db_size.to_i > available.to_i
+        Chef::Log.fatal("DB size is: #{db_size}, available space on #{path} is #{available}")
+        raise "Not enough space on #{path} for the Database dump!"
+      end
+    end
+  end
+
+  Chef::Log.info("dumping DB into #{dump_location}")
+
+  # Dump the database at the DB node
+  bash "dump database content" do
+    code <<-EOF
+      su - postgres -c 'pg_dumpall > #{dump_location}'
+    EOF
+  end
+
+elsif node["crowbar_wall"]["crowbar_db_shutdown"]
+
+  # Now, stop the database.
+  # In case of HA setup, also stop remaining pacemaker resources and HA services.
+  bash "stop the database" do
+    code <<-EOF
+      if test -e /usr/sbin/crm; then
+        crm resource stop postgresql
+        crm resource stop fs-postgresql
+        crm resource stop vip-admin-database-default-data
```

Comment:
Coming with a late review, but these lines are not fully correct:

- there can be a `g-postgresql` group to stop
- `vip-admin-database-default-data` is the name in your setup, but `default-data` comes from the name of the pacemaker proposal; so you should somehow find the right name
- you should stop the ms and drbd resources too (in case drbd is used)